src/main/java/com/google/cloud/tools/gradle/appengine/core/AppEngineCorePluginConfiguration.java

```diff
@@ -121,217 +123,164 @@ private void configureFactories() {
-                  });
-            });
+    createDeployTaskHelper(
+        DEPLOY_DOS_TASK_NAME, DeployDosTask.class, "Deploy Dos configuration", true);
   }
 
   private void createDeployIndexTask() {
-    injectGcloud(DeployIndexTask.class);
-    project
-        .getTasks()
-        .create(
-            DEPLOY_INDEX_TASK_NAME,
-            DeployIndexTask.class,
-            deployTask -> {
-              deployTask.setGroup(taskGroup);
-              deployTask.setDescription("Deploy Index configuration");
-
-              project.afterEvaluate(
-                  project -> {
-                    deployTask.setDeployExtension(deployExtension);
-                  });
-            });
+    createDeployTaskHelper(
+        DEPLOY_INDEX_TASK_NAME, DeployIndexTask.class, "Deploy Index configuration", true);
   }
 
   private void createDeployQueueTask() {
-    injectGcloud(DeployQueueTask.class);
-    project
-        .getTasks()
-        .create(
-            DEPLOY_QUEUE_TASK_NAME,
-            DeployQueueTask.class,
-            deployTask -> {
-              deployTask.setGroup(taskGroup);
-              deployTask.setDescription("Deploy Queue configuration");
-
-              project.afterEvaluate(
-                  project -> {
-                    deployTask.setDeployExtension(deployExtension);
-                  });
-            });
+    createDeployTaskHelper(
+        DEPLOY_QUEUE_TASK_NAME, DeployQueueTask.class, "Deploy Queue configuration", true);
   }
 
   private void createDeployAllTask() {
-    injectGcloud(DeployAllTask.class);
-    project
-        .getTasks()
-        .create(
-            DEPLOY_ALL_TASK_NAME,
-            DeployAllTask.class,
-            deployAllTask -> {
-              deployAllTask.setGroup(taskGroup);
-              deployAllTask.setDescription(
-                  "Deploy an App Engine application and all of its config files");
-
-              // deployConfig is set in AppEngineStandardPlugin and AppEngineAppYamlPlugin
-            });
+    createDeployTaskHelper(
+        DEPLOY_ALL_TASK_NAME,
+        DeployAllTask.class,
+        "Deploy an App Engine application and all of its config files",
+        false); // deployExtension is set in AppEngineStandardPlugin and AppEngineAppYamlPlugin
+  }
+
+  private void createDeployTaskHelper(
+      String taskName,
+      Class<? extends BaseDeployTask> taskClass,
+      String taskDescription,
+      boolean injectDeployExtension) {
+    TaskProvider<? extends BaseDeployTask> taskProvider =
+        project
+            .getTasks()
+            .register(
+                taskName,
+                taskClass,
+                task -> {
+                  task.setGroup(taskGroup);
+                  task.setDescription(taskDescription);
+                });
+
+    if (injectDeployExtension) {
+      project.afterEvaluate(
+          project -> taskProvider.configure(task -> task.setDeployExtension(deployExtension)));
+    }
```

Comment:
What is the purpose of `afterEvaluate` here?
It looks like `task.setDeployExtension(deployExtension)` can be called during task registration.

Reply:
I have the same question. (I have never worked on this repo, and I don't have a lot of Gradle knowledge.)

What I can see is that `deployExtension` is instantiated in the following way in `Plugin.apply()`.

First, a parent extension is created with `.create()`
```
    appengineExtension =
        project.getExtensions().create("appengine", AppEngineAppYamlExtension.class);
```
and then the sub-extension (`this` below is `appengineExtension` created above)
```
    deploy =
        ((ExtensionAware) this).getExtensions().create(DEPLOY_EXT, DeployExtension.class, project);
```
This `deploy` is what is called `deployExtension` in this class (if I am following the code flow correctly).

So the question is, do we still need to use `afterEvaluate()` when calling `setDeployExtension(deployExtension)`? I guess not, because it's just passing and setting a Java reference. But since I have never worked on this repo, I'm concerned there may be something that I'm not seeing. WDYT?

Reply:
>I guess not, because it's just passing and setting a Java reference

Indeed. If the reference is not changed, then it can be passed without `afterEvaluate`.

Reply:
I removed more `afterEvaluate` blocks in 1ca1a13 and 29190de.